transport-native-unix-common/src/main/java/io/netty/channel/unix/DirectIoByteBufPool.java

```diff
@@ -0,0 +1,1211 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.unix;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.DuplicatedByteBuf;
+import io.netty.buffer.ReadOnlyByteBuf;
+import io.netty.buffer.SlicedByteBuf;
+import io.netty.buffer.SwappedByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.util.ByteProcessor;
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.concurrent.FastThreadLocal;
+import io.netty.util.internal.MathUtil;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.UnstableApi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+import java.nio.charset.Charset;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+final class DirectIoByteBufPool {
+
+    private static final int MAX_CACHED_SIZE = SystemPropertyUtil.getInt(
+            "jdk.nio.maxCachedBufferSize", Integer.MAX_VALUE); // Use the same property name as the JDK.
+
+    // We need the Cleaner to ensure the memory will eventually be released as the user may not use a
+    // FastThreadLocalThread.
+    private static final ByteBufAllocator IO_BUFFER_ALLOCATOR = new UnpooledByteBufAllocator(true, true, false);
+
+    private static final FastThreadLocal<DirectIoByteBufPool> POOL = new FastThreadLocal<DirectIoByteBufPool>() {
+        @Override
+        protected DirectIoByteBufPool initialValue() {
+            return new DirectIoByteBufPool(Thread.currentThread());
+        }
+
+        @Override
+        protected void onRemoval(DirectIoByteBufPool value) {
+            value.freeAll();
+        }
+    };
+
+    private static final int ARRAY_LENGTH = Math.min(MathUtil.findNextPositivePowerOfTwo(Limits.IOV_MAX), 4096);
+    // We use a circular array to store the cached buffers.
+    private final PooledIoByteBuf[] buffers = new PooledIoByteBuf[ARRAY_LENGTH];
```

Comment:
I dont understand this class.   From my reading, it seems the sole purpose is to allocate the temporary IO buffers.  If that's true, the buffers allocated here are strictly limitted to being used in a single function call and then released.  That means that there is no need for an array of buffers, it should always be a single buffer of max size.  There should never be more than one, right?

Reply:
no that is not true...  as a user may write a combination of direct, heap buffers. For example something like:

```
writev(directBuffer, heapBuffer, directBuffer, heapBuffer, heapBuffer....);
```

Reply:
But that shouldn't matter.  You could allocate a single 32K(or however much) direct buf, and slice it for each of the writes.   That would avoid needing the array, altogether, and the scanning logic to find a correctly sized slice.   Also, the point still stands that these bufs are never cross thread, and are never under application control.

Reply:
But wouldn’t I need to scan all buffets then first to know how big the one buffer needs to be ? Also I think an array is better as it will reduce GC

Reply:
that said slicing out of one `ByteBuf` may waste less memory. Let me do some experiments...